### ISSUE TITLE
Is there any way to assign a custom declaration to a package?

### ISSUE THREAD

**Issue by atrauzzi** [Microsoft/TypeScript#14175]
So, before getting into the details, I just want to say that I suspect this question is easily misunderstood.  Also, that I feel like I have done at least some degree of research into my options, and either can't locate the correct guidance, it doesn't exist or what I'm looking to do can't be done (yet?).

The basic gist of it can be seen in this [Stack Overflow](http://stackoverflow.com/questions/42326246/how-do-i-reference-a-custom-declaration-in-my-project-for-a-module-im-importing) question.

The moderation pressure there may have started to confuse the question, so I'm going to try and state it simply here and hopefully the TS team can help me out :)

Basically I have a package ([postal](https://github.com/postaljs/postal.js)) which [I suspect has a few issues in its declaration file](https://github.com/DefinitelyTyped/DefinitelyTyped/issues/14717).

In the meantime however, I'd like to provide my own supplementary declarations for the package as part of my own project.  Unfortunately, I'm having a difficult time determining what the correct approach is to accomplish this - if it's even possible...

**Comment by mohsen1** [comment]
Take a look at MobX as an example package that provides its own typing:

https://github.com/mobxjs/mobx/blob/master/package.json#L6

**Comment by blakeembrey** [comment]
@atrauzzi There's two ways you can do it:

1. Write the declaration in "global" format (e.g. `declare module "x"`) and include it in your project - this is what Typings does
2. Since TypeScript 2.0, you can map module names to a definition using the `paths` option in `tsconfig.json`

**Comment by atrauzzi** [comment]
@mohsen1 - I'm not looking to include typings with my own package.  I'm looking to provide typings for a package that doesn't have them from within my own project.

Which FWIW is why I say this is easily misunderstood (but totally not your fault, cognitively TS is just very very complex).  ;)

**Comment by atrauzzi** [comment]
@blakeembrey -  Hmm, I think I understand your 1st option is suggesting.  I've got a `.d.ts` file in my project, but with the npm package installed, I get `Cannot find namespace "postal"`.

This is the `.d.ts` I've made:

```
declare module "postal" {

    export interface Configuration {

        SYSTEM_CHANNEL: string;
        DEFAULT_CHANNEL: string;
        resolver: Resolver;
    }

    export interface Resolver {

        compare(binding: string, topic: string, headerOptions: {}): boolean;
        reset(): void;
        purge(options?: {topic?: string, binding?: string, compact?: boolean}): void;
    }

    export interface Callback<T> {

        (data: T, envelope: Envelope<T>): void
    }

    export interface SubscriptionDefinition<T> {

        channel: string;
        topic: string;
        callback: Callback<T>;

        // after and before lack documentation

        constraint(predicateFn: (data: T, envelope: Envelope<T>) => boolean): SubscriptionDefinition<T>;
        constraints(predicateFns: ((data: T, envelope: Envelope<T>) => boolean)[]): SubscriptionDefinition<T>;
        context(theContext: any): SubscriptionDefinition<T>;
        debounce(interval: number): SubscriptionDefinition<T>;
        defer(): SubscriptionDefinition<T>;
        delay(waitTime: number): SubscriptionDefinition<T>;
        disposeAfter(maxCalls: number): SubscriptionDefinition<T>;
        distinct(): SubscriptionDefinition<T>;
        distinctUntilChanged(): SubscriptionDefinition<T>;
        logError(): SubscriptionDefinition<T>;
        once(): SubscriptionDefinition<T>;
        throttle(interval: number): SubscriptionDefinition<T>;
        subscribe(callback: Callback<T>): SubscriptionDefinition<T>;
        unsubscribe(): void;
    }

    export interface Envelope<T> {

        topic: string;
        data?: T;

        /*Uses DEFAULT_CHANNEL if no channel is provided*/
        channel?: string;

        timeStamp?: string;
    }


    export interface ChannelDefinition<T> {

        subscribe(topic: string, callback: Callback<T>): SubscriptionDefinition<T>;

        publish(topic: string, data?: T): void;

        channel: string;
    }

    export interface Postal {

        subscriptions: {};
        wireTaps: Callback<any>[];

        addWireTap(callback: Callback<any>): () => void;

        channel<T>(name?: string): ChannelDefinition<T>;

        getSubscribersFor(): SubscriptionDefinition<any>[];
        getSubscribersFor(options: {channel?: string, topic?: string, context?: any}): SubscriptionDefinition<any>[];
        getSubscribersFor(predicateFn: (sub: SubscriptionDefinition<any>) => boolean): SubscriptionDefinition<any>[];

        publish(envelope: Envelope<any>): void;

        reset(): void;

        subscribe(options: {channel?: string, topic: string, callback: Callback<any>}): SubscriptionDefinition<any>;
        unsubscribe(sub: SubscriptionDefinition<any>): void;
        unsubscribeFor(): void;
        unsubscribeFor(options: {channel?: string, topic?: string, context?: any}): void;

        configuration: Configuration;
    }

    const postal: Postal;
    export default postal;
}

```

Compared to the one over at DefinitelyTyped, I'm just trying to export the contracts, and also remove the `I` in front of everything as internally that's not how postal names things.

**Comment by blakeembrey** [comment]
@atrauzzi Make sure your custom `.d.ts` file is included in the project somehow. What is the import your attempting to do? Also, please note that `export default` and `export =` mean very different things - most likely you still want `export =` over `export default` (`export =` is synonymous with `module.exports =`). Using `export =` means you cannot use `export` elsewhere and you'll usually follow a pattern that looks more like https://github.com/typings/typings/blob/master/docs/examples.md#utility-library.

**Comment by atrauzzi** [comment]
@blakeembrey - Alright, so going with the limitations of `export =`, which is what the current typings are using, how do I get the type information exported correctly, without clobbering the actual package's default export?  Is this even a default export anymore?  Definitely confused between the two.

**Comment by blakeembrey** [comment]
@atrauzzi See the link above for some examples. You can just do all the work in a namespace and then export the namespace using `export =`. When you import it, you'd use `import x = require('x')`. This is TypeScript's support for the CommonJS `module.exports =` style of coding.

However, if the module isn't exporting a function and only uses the CommonJS `exports.foo =` style, you can write it in the normal ES6 style. You wouldn't need either of `export default` or `export =` in that case. The difference between `export =` and `export default` really only matters in transpilation in this case - `export default` and `import foo from` are treated like `exports.default =` and `var x = require().default` in CommonJS while `export =` and `import x = require()` is treated like `module.exports =` and `var x = require('')` in CommonJS.

It does look like Postal itself is just an object exported, so I wrote this based on your previously posted code:

```ts
declare module "postal" {
    export interface Configuration {
        SYSTEM_CHANNEL: string;
        DEFAULT_CHANNEL: string;
        resolver: Resolver;
    }

    export interface Resolver {
        compare(binding: string, topic: string, headerOptions: {}): boolean;
        reset(): void;
        purge(options?: {topic?: string, binding?: string, compact?: boolean}): void;
    }

    export interface Callback<T> {
        (data: T, envelope: Envelope<T>): void
    }

    export interface SubscriptionDefinition<T> {

        channel: string;
        topic: string;
        callback: Callback<T>;

        // after and before lack documentation

        constraint(predicateFn: (data: T, envelope: Envelope<T>) => boolean): SubscriptionDefinition<T>;
        constraints(predicateFns: ((data: T, envelope: Envelope<T>) => boolean)[]): SubscriptionDefinition<T>;
        context(theContext: any): SubscriptionDefinition<T>;
        debounce(interval: number): SubscriptionDefinition<T>;
        defer(): SubscriptionDefinition<T>;
        delay(waitTime: number): SubscriptionDefinition<T>;
        disposeAfter(maxCalls: number): SubscriptionDefinition<T>;
        distinct(): SubscriptionDefinition<T>;
        distinctUntilChanged(): SubscriptionDefinition<T>;
        logError(): SubscriptionDefinition<T>;
        once(): SubscriptionDefinition<T>;
        throttle(interval: number): SubscriptionDefinition<T>;
        subscribe(callback: Callback<T>): SubscriptionDefinition<T>;
        unsubscribe(): void;
    }

    export interface Envelope<T> {

        topic: string;
        data?: T;

        /*Uses DEFAULT_CHANNEL if no channel is provided*/
        channel?: string;

        timeStamp?: string;
    }


    export interface ChannelDefinition<T> {

        subscribe(topic: string, callback: Callback<T>): SubscriptionDefinition<T>;

        publish(topic: string, data?: T): void;

        channel: string;
    }

    export var subscriptions: {};
    export var wireTaps: Callback<any>[];
    export var configuration: Configuration;

    export function addWireTap(callback: Callback<any>): () => void;

    export function channel<T>(name?: string): ChannelDefinition<T>;

    export function getSubscribersFor(): SubscriptionDefinition<any>[];
    export function getSubscribersFor(options: {channel?: string, topic?: string, context?: any}): SubscriptionDefinition<any>[];
    export function getSubscribersFor(predicateFn: (sub: SubscriptionDefinition<any>) => boolean): SubscriptionDefinition<any>[];

    export function publish(envelope: Envelope<any>): void;

    export function reset(): void;

    export function subscribe(options: {channel?: string, topic: string, callback: Callback<any>}): SubscriptionDefinition<any>;
    export function unsubscribe(sub: SubscriptionDefinition<any>): void;
    export function unsubscribeFor(): void;
    export function unsubscribeFor(options: {channel?: string, topic?: string, context?: any}): void;
}
```

**Comment by atrauzzi** [comment]
That list of various export approaches might take a bit for me to internalize an understanding of...  But I think I get some of them.  Maybe.

I know right now I'm successfully importing the js via `import * as postal from "postal"`.  I understand there have been some changes in TS2.2 on this front?  I'm hoping I don't have to write the gross `=` style of import as it's terribly inconsistent.

**Comment by blakeembrey** [comment]
That syntax works fine. It's just saying import all the properties as an object basically. You don't need the `=` style in this case, since there's no function exported on `module.exports`. You really only need it when it's the `module.exports =` style of code which is common in node.js - e.g. `express` and other small utility libraries.

**Comment by atrauzzi** [comment]
Great, I think I've picked up a few useful pieces of info here.  I'll keep going forward with it for now.
